[LTB-16] add support for sum types

pasqu4le opened the pull request:
This PR adds structures in `loot-config` similar to sum-types with the intention of supersede them in config contexts.

The changes in the README, and the comments added should explain the basics of how it works, I'm of course available to provide further explanations.

I uploaded this so reviews can start, but it is labeled as a WIP because I'd like to try to improve it some more or some things may need to be changed heavily, in particular:
- I'd like for the type system to be more aware of the fact that `tree`s only contain `branches`, this would make it more robust and simplify some code, by removing unnecessary checks, but I'm not sure how (or if, really) it can be done
- or ... not. One thing worth considering IMHO is that, somewhat differently from sum-types, `tree`s could also have `option`s/`sub`s that are common and will be used in any case *and* `branch`es that only have the differences
- Some more "smart" lenses could really come in handy, to more easily set/use `branch`es (in particular the selected one) especially after they have been `finalise`d and are inside `Maybe`s
- I tested this and it works as expected, but I did not add proper tests for this yet because it may need to be modified a lot. Old tests still work because the changes are only incremental.

flyingleafe commented on code/config/lib/Loot/Config/Record.hs:
```diff
@@ -83,20 +107,45 @@ type family Item' (k :: ConfigKind) (i :: ItemKind) where
+    Item' 'Partial (l ::+ is) = ConfigRec 'Partial (SumSelection l : ToBranches is)
+    Item' 'Final   (l ::+ is) = ConfigRec 'Final   (SumSelection l : ToBranches is)
+    Item' 'Partial (l ::- is) = ConfigRec 'Partial is
+    Item' 'Final   (l ::- is) = Maybe (ConfigRec 'Final is)
```
I started to write a long comment about how it would be great to just use an actual sum type instead of a product of `Maybe`s with saved `SumSelection` in final config, but then I realized that `vinyl` provides _only_ sum types (`Rec`s), and migrating the whole `loot-config` to `ntype` (https://github.com/int-index/ntype) would be quite an ordeal - almost the same ordeal as defining N-ary abstract sum type from scratch. So okay, this is a fine middle-ground.

pasqu4le replied:
> `vinyl` provides _only_ sum types (`Rec`s)

you mean only *product* types right?

Anyway another advantage of doing it this way is that it is not disruptive and whoever was using `loot-config` before should not have any problem updating.

flyingleafe replied:
>you mean only product types right?
Yes, that was a mistake.

flyingleafe commented on code/config/lib/Loot/Config/Record.hs:
```diff
@@ -133,7 +202,17 @@ type family SubRecsConstrained c k is :: Constraint where
-
+    SubRecsConstrained c k ((l ::+ us) ': is) =
+        ( c (ConfigRec k (SumSelection l : ToBranches us))
+        , SubRecsConstrained c k (SumSelection l : ToBranches us)
```
Including `SumSelection` in this line is simply useless, it will be skipped by second pattern-match anyway.

pasqu4le replied:
Indeed, has been removed

flyingleafe commented on code/config/lib/Loot/Config/Record.hs:
```diff
@@ -133,7 +202,17 @@ type family SubRecsConstrained c k is :: Constraint where
+        )
+    SubRecsConstrained c k ((_ ::- us) ': is) =
+        ( c (ConfigRec k us)
+        , c (Maybe (ConfigRec k us))
```
You only need `c (ConfigRec k us)` if `k ~ 'Partial` and `c (Maybe (ConfigRec k us))` if `k ~ 'Final`. This should be split in two pattern matches, matching also by `k`.

pasqu4le replied:
It is right that I only need `c (ConfigRec k us)` if `k ~ 'Partial` (so I did split it in two pattern matches), but I don't think that `c (Maybe (ConfigRec k us))` is enough when `k ~ 'Final`.

Take the `Show` instance for `Maybe`: it is `Show a => Show (Maybe a)`, it would be enough if we only use Show on a `Maybe (ConfigRec k us)`, but the type system cannot guarantee that `ConfigRec k us` is part of `Show` as well.
We actually use this (in the `Show (Item k i)` instance definition), so we actually need to impose `Show` on `c (ConfigRec k us)` as well.

Also, this case seems to be very common, so I left it for now, tell me what you think.

flyingleafe replied:
@pasqu4le Actually, I just tried to remove constraint `c (Maybe (ConfigRes k us))` altogether, leaving only this:
```
    SubRecsConstrained c k ((_ ::- us) ': is) =
        ( c (ConfigRec k us)
        , SubRecsConstrained c k us
        , SubRecsConstrained c k is
        )
```
and it works just fine, analogously with how `c v` without `c (Maybe v)` constraint works just fine for any kind of configs in `ValuesConstrained`.

pasqu4le replied:
You are right, the constraint on the`Maybe` never get used and it's not necessary at all. Removed.

flyingleafe commented on code/config/lib/Loot/Config/Record.hs:
```diff
@@ -106,12 +157,22 @@ itemOptionLabel _ = symbolVal (Proxy :: Proxy l)
+
+-- | Internal helper used to get the name of a branch given the branch.
+itemBranchLabel :: forall k l is. KnownSymbol l => Item k (l ::- is) -> String
+itemBranchLabel _ = symbolVal (Proxy :: Proxy l)
```
I'm thinking of replacing this zoo with
```
type family ItemLabel (i :: ItemKind) where
    ItemLabel (l ::: _) = l
    ...
    ItemLabel (l ::- _) = l

itemLabel :: forall k i. KnownSymbol (ItemLabel i) => Item k i -> String
itemLabel _ = symbolVal (Proxy :: Proxy (ItemLabel i))
```

pasqu4le replied:
Done

flyingleafe commented on code/config/lib/Loot/Config/Yaml.hs:
```diff
@@ -54,6 +57,33 @@ instance
+    configParser = (:&) <$> fmap ItemSumP parseSum <*> configParser
+      where
+        parseSum :: Parse e (ConfigRec 'Partial (SumSelection l : ToBranches us))
+        parseSum = keyOrDefault (fromString $ symbolVal (Proxy :: Proxy l)) mempty configParser
```
Copy-paste worth moving to a separate function.

pasqu4le replied:
Done this too

pasqu4le commented on code/config/lib/Loot/Config/Record.hs:
```diff
@@ -121,6 +180,14 @@ type family ValuesConstrained c is :: Constraint where
         , ValuesConstrained c is
         )
+    ValuesConstrained c ((l ::+ us) ': is) =
+        ( ValuesConstrained c (SumSelection l : ToBranches us)
```
I agree with you, I remember adding that constraint because something was not compiling and I thought that it made sense: `SumSelection l` is effectively another `option` so that should be checked as well, right?

Apparently not, I removed it now and tested it a bit, it all seems to compile and work as expected.

pasqu4le commented:
@flyingleafe I addressed your comments and I also thought more about further improvements/changes.

I don't think the idea of "smarter" lenses can be pushed more, the only think I made a `selection` lens, because the types are aware of the `SumSelection l`, but I wasn't able to do anything related to branches, because that would need some function work that the type families cannot be aware of.

But, I think we should allow non-`branch` item inside of `tree`s, we would:
- remove the `ToBranches` type family: the conversion that it does is somewhat arbitrary, and it's not useful later on, because AFAIU the type system cannot be forced to know that the result of `ToBranches` is always a `branch`
- simplify all the constraints, and most importantly remove the need for the `finalizeBranches` and `finaliseDeferredUnsafeBranches` functions:  right now this are poorly made because we cannot prove that the "catastrophic" case is impossible, but necessary because we need it to finalize only the selected branch
- merge the ability to "look for a branch" inside `finalize'` and get rid of the current useless behavior, that is also quite arbitrary

In the end this would greatly improve the code IMHO, still allow to only have branches in a tree (one just needs to use `::-`) and also allow to define part that would be common to all branches in the tree itself (with `option` and `sub`).I hope you agree with me and that we go for it.

This also should not take much time at all. After this only test remains to be updated and this is done.

pasqu4le commented:
@flyingleafe everything should be done, all your comments have been addressed and the tests have been updated.
